src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -550,6 +579,61 @@ Node *RegionNode::Ideal(PhaseGVN *phase, bool can_reshape) {
           }
         }
       }
+    } else {
+      // in passive materialization, we create a control node which covers multiple inputs of a region node.
+      //
+      // +-----------+   +--------+
+      // | CFG node  |   |cfg n   |
+      // +-----------+   +--------+
+      //      |  \  \     /
+      //      |   |  |   |
+      // +----------------+
+      // | Region         |
+      // +----------------+
+      //
+      // we need to normalize the region node, and its phi nodes.
+      // +-----------+  +-----+
+      // | CFG node  |  |cfg n|
+      // +-----------+  +-----+
+      //      |        /
+      //      |       |
+      // +-------------+
+      // | Region      | (inputs subtracts 2)
+      // +-------------+
+      //
+      for (uint i = 1; i < req()-1; i++) {
+        Node* c1 = in(i);
+        if (c1 == nullptr || c1->is_top()) continue;
+
+        GrowableArray<uint> paths;
+        for (uint j=i+1; j<req(); ++j) {
+          if (in(j) == c1) {
+            paths.push(j);
+          }
+        }
+
+        if (paths.length() > 0) {
+          PhaseIterGVN* igvn = phase->is_IterGVN();
+
+          for (DUIterator_Fast kmax, k = fast_outs(kmax); k < kmax; k++) {
+            Node* phi = fast_out(k);
+
+            if (phi != nullptr && phi->is_Phi() && phi->req() == req()) {
+              NodeChanger nc(igvn, phi);
+              for (int j = paths.length()-1; j>=0; --j) {
+                nc.del_req(paths.at(j));
+              }
+              igvn->add_users_to_worklist(phi);
+            }
+          }
+
+          NodeChanger nc(igvn, this);
+          for (int j = paths.length()-1; j>=0; --j) {
+            nc.del_req(paths.at(j));
+          }
+          return this;
```

Comment:
Is this an early return? We can only remove one set of dupllicates. We` can have multiple sets of duplicates. For example

```
200 Region === [200, 100, 100, 101, 101]
```

I don't think PEA will do this, but maybe it would make sense for common RegionNode code.

Reply:
this is done by PhaseGVN. transform() keep calling ideal() until it turns nullptr. 
  Node* k = n;
  Node* i = apply_ideal(k, /*can_reshape=*/false);
  NOT_PRODUCT(uint loop_count = 1;)
  while (i != nullptr) {
    assert(i->_idx >= k->_idx, "Idealize should return new nodes, use Identity to return old nodes" );
    k = i;
    i = apply_ideal(k, /*can_reshape=*/false);
    NOT_PRODUCT(loop_count++;)
  }
```